src/kvstore/raftex/RaftPart.cpp

```diff
@@ -415,6 +419,82 @@ void RaftPart::commitTransLeader(const HostAddr& target) {
     }
 }
 
+void RaftPart::updateQuorum() {
+    CHECK(!raftLock_.try_lock());
+    int32_t total = 0;
+    for (auto& h : hosts_) {
+        if (!h->isLearner()) {
+            total++;
+        }
+    }
+    quorum_ = (total + 1) / 2;
+}
+
+void RaftPart::addPeer(const HostAddr& peer) {
+    CHECK(!raftLock_.try_lock());
+    if (peer == addr_) {
+        LOG(INFO) << idStr_ << "I am already in the raft group!";
+        return;
+    }
+    auto it = std::find_if(hosts_.begin(), hosts_.end(), [&peer] (const auto& h) {
+                return h->address() == peer;
+            });
+    if (it == hosts_.end()) {
+        hosts_.emplace_back(std::make_shared<Host>(peer, shared_from_this()));
+        updateQuorum();
+        LOG(INFO) << idStr_ << "Add peer " << peer;
+    } else {
+        if ((*it)->isLearner()) {
```

Comment:
When do we addPeer the second times? When we don't WAIT_SNAPSHOT?

Reply:
Because when we add one new host, we hope it will be added as learner,  after catching up data,  we promote it to be follower.

Reply:
Yeah, that is exactly what I mean. I have not see it this morning when it mixes up with snapshot.